File: forge/src/main/java/juuxel/adorn/lib/AdornCapabilities.java
package juuxel.adorn.lib;

import juuxel.adorn.block.AdornBlockEntities;
import juuxel.adorn.block.entity.BlockEntityWithFluidTank;
import juuxel.adorn.block.variant.BlockKind;
import juuxel.adorn.block.variant.BlockVariantSets;
import net.minecraft.core.Direction;
import net.minecraft.world.Container;
import net.minecraft.world.WorldlyContainer;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.neoforged.neoforge.capabilities.Capabilities;
import net.neoforged.neoforge.capabilities.IBlockCapabilityProvider;
import net.neoforged.neoforge.capabilities.ICapabilityProvider;
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent;
import net.neoforged.neoforge.transfer.ResourceHandler;
import net.neoforged.neoforge.transfer.fluid.FluidResource;
import net.neoforged.neoforge.transfer.item.ItemResource;
import net.neoforged.neoforge.transfer.item.VanillaContainerWrapper;
import net.neoforged.neoforge.transfer.item.WorldlyContainerWrapper;
import org.jspecify.annotations.Nullable;

public final class AdornCapabilities {
    private static final IBlockCapabilityProvider<ResourceHandler<ItemResource>, @Nullable Direction> INVENTORY_WRAPPER_FOR_BLOCK =
        (world, pos, state, blockEntity, side) -> blockEntity instanceof Container inventory ? getInventoryWrapper(inventory, side) : null;
    private static final ICapabilityProvider<BlockEntity, @Nullable Direction, ResourceHandler<ItemResource>> INVENTORY_WRAPPER_FOR_BLOCK_ENTITY =
        (blockEntity, side) -> blockEntity instanceof Container inventory ? getInventoryWrapper(inventory, side) : null;
    private static final IBlockCapabilityProvider<ResourceHandler<FluidResource>, @Nullable Direction> FLUID_TANK_FOR_BLOCK =
        (world, pos, state, blockEntity, side) -> blockEntity instanceof BlockEntityWithFluidTank withTank ? withTank.getTank() : null;
    private static final ICapabilityProvider<BlockEntity, @Nullable Direction, ResourceHandler<FluidResource>> FLUID_TANK_FOR_BLOCK_ENTITY =
        (blockEntity, side) -> blockEntity instanceof BlockEntityWithFluidTank withTank ? withTank.getTank() : null;

    public static void register(RegisterCapabilitiesEvent event) {
        event.registerBlockEntity(Capabilities.Item.BLOCK, AdornBlockEntities.BREWER.get(), INVENTORY_WRAPPER_FOR_BLOCK_ENTITY);

        var containerBlockKinds = new BlockKind[] {
            BlockKind.DRAWER,
            BlockKind.KITCHEN_CUPBOARD,
            BlockKind.SHELF,
        };

        for (var kind : containerBlockKinds) {
            for (var block : BlockVariantSets.get(kind)) {
                event.registerBlock(Capabilities.Item.BLOCK, INVENTORY_WRAPPER_FOR_BLOCK, block.get());
            }
        }

        event.registerBlockEntity(Capabilities.Fluid.BLOCK, AdornBlockEntities.BREWER.get(), FLUID_TANK_FOR_BLOCK_ENTITY);

        for (var kitchenSink : BlockVariantSets.get(BlockKind.KITCHEN_SINK)) {
            event.registerBlock(Capabilities.Fluid.BLOCK, FLUID_TANK_FOR_BLOCK, kitchenSink.get());
        }
    }

    private static ResourceHandler<ItemResource> getInventoryWrapper(Container inventory, @Nullable Direction side) {
        return side != null && inventory instanceof WorldlyContainer sided ? new WorldlyContainerWrapper(sided, side) : VanillaContainerWrapper.of(inventory);
    }
}

File: forge/src/main/kotlin/juuxel/adorn/platform/forge/AdornCapabilities.kt
package juuxel.adorn.platform.forge

import juuxel.adorn.block.AdornBlockEntities
import juuxel.adorn.block.variant.BlockKind
import juuxel.adorn.block.variant.BlockVariantSets
import juuxel.adorn.platform.forge.block.entity.BlockEntityWithFluidTank
import net.minecraft.block.entity.BlockEntity
import net.minecraft.inventory.Inventory
import net.minecraft.inventory.SidedInventory
import net.minecraft.util.math.Direction
import net.neoforged.bus.api.SubscribeEvent
import net.neoforged.neoforge.capabilities.Capabilities
import net.neoforged.neoforge.capabilities.IBlockCapabilityProvider
import net.neoforged.neoforge.capabilities.ICapabilityProvider
import net.neoforged.neoforge.capabilities.RegisterCapabilitiesEvent
import net.neoforged.neoforge.fluids.capability.IFluidHandler
import net.neoforged.neoforge.items.IItemHandler
import net.neoforged.neoforge.items.wrapper.InvWrapper
import net.neoforged.neoforge.items.wrapper.SidedInvWrapper

object AdornCapabilities {
    private val INVENTORY_WRAPPER_FOR_BLOCK: IBlockCapabilityProvider<IItemHandler, Direction?> =
        IBlockCapabilityProvider { _, _, _, blockEntity, side ->
            if (blockEntity is Inventory) {
                getInventoryWrapper(blockEntity, side)
            } else {
                null
            }
        }

    private val INVENTORY_WRAPPER_FOR_BLOCK_ENTITY: ICapabilityProvider<BlockEntity, Direction?, IItemHandler> =
        ICapabilityProvider { blockEntity, side ->
            if (blockEntity is Inventory) {
                getInventoryWrapper(blockEntity, side)
            } else {
                null
            }
        }

    private val FLUID_TANK_FOR_BLOCK: IBlockCapabilityProvider<IFluidHandler, Direction?> =
        IBlockCapabilityProvider { _, _, _, blockEntity, side ->
            if (blockEntity is BlockEntityWithFluidTank) {
                blockEntity.tank
            } else {
                null
            }
        }

    private val FLUID_TANK_FOR_BLOCK_ENTITY: ICapabilityProvider<BlockEntity, Direction?, IFluidHandler> =
        ICapabilityProvider { blockEntity, _ ->
            if (blockEntity is BlockEntityWithFluidTank) {
                blockEntity.tank
            } else {
                null
            }
        }

    @SubscribeEvent
    fun register(event: RegisterCapabilitiesEvent) {
        event.registerBlockEntity(Capabilities.ItemHandler.BLOCK, AdornBlockEntities.BREWER, INVENTORY_WRAPPER_FOR_BLOCK_ENTITY)

        val containerBlockKinds = listOf(
            BlockKind.DRAWER,
            BlockKind.KITCHEN_CUPBOARD,
            BlockKind.SHELF,
        )

        for (kind in containerBlockKinds) {
            for (block in BlockVariantSets.get(kind)) {
                event.registerBlock(Capabilities.ItemHandler.BLOCK, INVENTORY_WRAPPER_FOR_BLOCK, block.get())
            }
        }

        event.registerBlockEntity(Capabilities.FluidHandler.BLOCK, AdornBlockEntities.BREWER, FLUID_TANK_FOR_BLOCK_ENTITY)

        for (kitchenSink in BlockVariantSets.get(BlockKind.KITCHEN_SINK)) {
            event.registerBlock(Capabilities.FluidHandler.BLOCK, FLUID_TANK_FOR_BLOCK, kitchenSink.get())
        }
    }

    private fun getInventoryWrapper(inventory: Inventory, side: Direction?): IItemHandler =
        if (side != null && inventory is SidedInventory) {
            SidedInvWrapper(inventory, side)
        } else {
            InvWrapper(inventory)
        }
}
